### PR TITLE
Update CPG Actions

### PR DESCRIPTION
**Pull request by blangley28**
Fix the CPG Actions in the CPGExporter to be inline with [this example](https://jira.mitre.org/secure/attachment/42600/Early-Stage-HER-2-CPG.json)

Big changes are:
* Pathways are mapped to a RecommendationDefinition
* Actions in the Pathway are repeated (but share ActivityDefinitions)
* To achieve this the pathway is "cleaned" prior to exporting - this copies sections of the pathway so each node has only one parent. I left three new pathways in the builder with suffix "CLEAN" to illustrate what these pathways look like (will delete before merging)
* The exporter then goes through a modified BFS (to ensure parent is always visited) and converts each node

I think everything on the actions should be correct (sub actions, conditions, related actions, and definitionCanonical) but I'm looking for more sets of eyes on this. 

I tried to leave comments for chunks of code which may not be obviously understandable. If you see anything which doesn't make sense please leave a comment - I was already confusing myself after the long weekend off so we should emphasis readability and understandability of this code.

### CODE REVIEW

File: src/utils/cpg.ts [hunk 1]
```diff
@@ -18,113 +30,354 @@ const BUNDLE_TRANSACTION = R4.BundleTypeKind._transaction;
 const BUNDLE_PUT = R4.Bundle_RequestMethodKind._put;
 const CONDITION_APPLICABILITY = R4.PlanDefinition_ConditionKindKind._applicability; // eslint-disable-line
 const EXPRESSION_CQL = R4.ExpressionLanguageKind._textCql;
+const CHILD_RELATIONSHIP = PlanDefinition_RelatedActionRelationshipKind._beforeStart; // eslint-disable-line
```
[the remark this answers is not in the document]
Comment (dehall) [reply src/utils/cpg.ts:33]: Either way it makes sense in the resulting JSON

**Review by dehall — COMMENTED** [review]
I haven't tested anything especially complex but I have done some basic testing and overall this looks good. A few minor issues to resolve, plus some suggestions for spots that might benefit from extra clarity/comments

File: src/utils/cpg.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (blangley28) [reply src/utils/cpg.ts:33]: Yea it's read as "this node" {relationship kind} "related id node". So the before start relationship means complete this current action before starting this related action - which means it the relationship to identify the child.

File: src/utils/cpg.ts [hunk 2]
```diff
@@ -133,120 +386,63 @@ function createBundleEntry(resource: PlanDefinition | ActivityDefinition | Libra
...
-          cpgStrategyAction.condition = cpgStrategyAction.condition || [];
-          cpgStrategyAction.condition?.push(condition);
+export function cleanPathway(nodes: NodeObj): NodeObj {
+  const newNodes = nodes;
+  Object.keys(nodes).forEach(key => {
+    const allTransitions = findAllTransistions(nodes, key);
+
+    // Remove multiple transitions to node by copying subpathway
+    if (allTransitions.length > 1) {
+      allTransitions.shift(); // Remove one transition to maintain the original subpathway
```
[the remark this answers is not in the document]
Comment (blangley28) [reply src/utils/cpg.ts:432]: Yes your understanding of "maintain the original subpathway" is correct. By subpathway I mean the pathway rooted at the current node. What do you suggest calling it?

File: src/utils/cpg.ts [hunk 3]
```diff
@@ -18,113 +30,354 @@ const BUNDLE_TRANSACTION = R4.BundleTypeKind._transaction;
...
+      const parentAction = findAction(planDefinition.action, parentKey);
+      if (!parentAction) return;
+      else if (parentAction.action?.length) parentAction.action.push(action);
+      else parentAction.action = [action];
+    } else planDefinition.action.push(action);
+  }
+}
+
+function formatUrl(id: string): string {
+  return `http://pathway.com/${id}`;
```
[the remark this answers is not in the document]
Comment (blangley28) [reply src/utils/cpg.ts]: Does the FY21 budget include purchasing pathway.com or pathways.com

File: src/utils/cpg.ts [hunk 4]
```diff
@@ -18,113 +30,354 @@ const BUNDLE_TRANSACTION = R4.BundleTypeKind._transaction;
...
+      action.resource.resourceType === 'Procedure'
+        ? 'ServiceRequest'
+        : action.resource.resourceType;
+    const codeableConcept =
       action.resource.resourceType === 'MedicationRequest'
         ? action.resource.medicationCodeableConcept
-        : action.resource.code
-  };
+        : action.resource.code;
+    const coding = codeableConcept.coding[0];
```
[the remark this answers is not in the document]
Comment (blangley28) [reply src/utils/cpg.ts:129]: Updated so CarePlans will have a fake codebaleconcept now

**Review by dehall — APPROVED** [review]
I'm good with this as-is but I had one other thought for improvement. Right now the description of actions is based on just the current node (`Action for {label}`), but since there can be multiple actions generated from the same original node, it's not immediately obvious what the difference is (at least not without tracing through the IDs). If there's an easy way to generate a simple "path name" that includes the parent name as well that would make visual inspection of the FHIR a lot easier to follow. But if it's not easy it's not a blocker

**Comment by blangley28** [comment]
@dehall added `(Key: ${key}) - Parent Node Key: ${parentKey}` to each action description

**Review by radamson — APPROVED** [review]
Needs a rebase, but looks good and the shared ActivityDefinitions are 👌 .

**Pull request merged**